aten/src/ATen/native/cpu/SampledAddmmKernel.cpp

```diff
@@ -0,0 +1,94 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/core/Tensor.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/cpu/vec/vec.h>
+#include <ATen/native/cpu/SampledAddmmKernel.h>
+#include <ATen/native/cpu/utils.h>
+#include <c10/util/irange.h>
+
+namespace at { namespace native {
+
+namespace {
+
+template <typename scalar_t, typename index_t>
+void sampled_addmm_sparse_csr_kernel_impl(
+    const Tensor& mat1,
+    const Tensor& mat2,
+    const Scalar& beta,
+    const Scalar& alpha,
+    const Tensor& result) {
+
+  int64_t nnz = result._nnz();
+
+  auto beta_ = beta.to<scalar_t>();
+  auto alpha_ = alpha.to<scalar_t>();
+
+  scalar_t* mat1_data = mat1.data_ptr<scalar_t>();
+  scalar_t* mat2_data = mat2.data_ptr<scalar_t>();
+  scalar_t* value_data = result.values().data_ptr<scalar_t>();
+  index_t* crow_data = result.crow_indices().data_ptr<index_t>();
+  index_t* col_data = result.col_indices().data_ptr<index_t>();
+
+  // mat1: {B, M, K}
+  // mat2: {B, N, K}
+  // crow: {B, M + 1}
+  // col, values: {B, nnz}
+  int64_t M = mat1.size(-2);
+  int64_t K = mat1.size(-1);
+  int64_t N = mat2.size(-2);
+  int64_t B = mat1.numel() / M / K;
+
+  // usually, collapse B and M is a better option,
+  // but for most commonly used case (mat1 and mat2 is 2d tensor), B = 1,
+  // so choose to parallel on M so that we don't have to handle b offset frequently
+  using Vec = vec::Vectorized<scalar_t>;
+  utils::parallel_sparse_csr(crow_data, M, nnz, [&](int64_t begin, int64_t end) {
+    for (const auto b : c10::irange(B)) {
+      index_t* crow_ptr = crow_data + b * (M + 1);
+      index_t* col_ptr = col_data + b * nnz;
+      scalar_t* value_ptr = value_data + b * nnz;
```

Comment:
https://github.com/pytorch/pytorch/pull/91243 added support for CSR indices contiguity per batch and non-contiguous values. Could we use an accessor here so that non-contiguous crow/col_indices and values in `result` could be supported?

Reply:
In order to accomplish that one could use a TI interface to ease batch advancement. Looks like `cpu_kernel_vec` could be used here.

Reply:
OK I see, with #91243 merged, we can't expect `values`, `crow` and `col` to be always contiguous. I have modified this PR to use `accessor` instead raw pointers.

Using TI for this one is not a good idea: if i use TI, it will be shape of {B, M} and squash {K}. The parallelization will be evenly done on {B, M} which is not the best choice from performance perspective. The reason is that on real GNN datasets, the edges each node have may vary quite a lot, consider the following example: say we have 10 nodes, and node0 has 10 edges, while node1~9 has only 1 edge, and we have 2 threads. TI will end up with `thread0: node0~4` and `thread1: 5~9`. The partition in this PR will have `thread0: node0` and `thread1:1~9`.

Reply:
Sorry, I am not sure I understand. It is possible to restride inputs whichever way it is beneficial for balanced workload (setting strides to zero and the corresponding dim sizes to suitable values). Given that `reshape` is used to squash batch dims, it is not that much different from the previous solution with calls to `contiguous` prior to kernel runs.